### PR TITLE
Redefine iree_ext.scatter for having index depth information.

### PR DESCRIPTION
**Pull request by hanhanW**
The index depth could be others than one. E.g.,

```
%0 = "mhlo.scatter"(%arg0, %arg1, %arg2) ( {
^bb0(%arg3: tensor<f32>, %arg4: tensor<f32>):  // no predecessors
  "mhlo.return"(%arg4) : (tensor<f32>) -> ()
}) {indices_are_sorted = false,
    scatter_dimension_numbers = {
      ...
    },
    unique_indices = false
} : (tensor<1x24x512xf32>, tensor<1x2xi32>, tensor<1x512xf32>)
  -> tensor<1x24x512xf32>
}
```

This is a step toward https://github.com/google/iree/issues/6403

### CODE REVIEW

File: iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 1]
```diff
@@ -46,6 +46,21 @@ def LinalgExt_ScatterOp : LinalgExt_Op<"scatter"> {
     `updates` (and `original`). The first argument correspond the
     value to be updated (i.e. from `updates`), and the second the
     current value (i.e. value from `original`).
+
+    The `indices` is a 2D tensor/memref type. The first dim is the number of
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:50]: Is it worth putting this in the `indices` operand. This seems totally unrelated to indices. If so I would make this a separate operand by itself.

Comment (MaheshRavishankar) [reply iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:50]: Also if this has to be static, then might as well make this a IntegerAttr.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 1 above)
Comment (hanhanW) [reply iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:50]: I'm really not sure how to not put it in the indices and how to just make it a IntegerAttr. Say that the indices is `AxB`. This means that there are `A` updates, (which can be treated as batch size), and each update *needs* `B` indices. Because we need `B` indices to access the `original`, (where `original` has `B+` rank).

`B` must be static, because we do know the rank of `original`. `B` + `dim(update) - 1` is exactly the rank of `original`. If it is not static, we can infer the value.

In the old definition, I think you we force the `orignial` rank being `update` rank plus one. Because the index depth is always a scalar, ie, `1`.

```
func @scatter_tensor_static(
    %original: tensor<128x3xf32>, %indices: tensor<48xi32>,
    %update: tensor<48x3xf32>) -> tensor<128x3xf32> {
  %0 = linalg_ext.scatter
    ins(%update, %indices : tensor<48x3xf32>, tensor<48xi32>)
    outs(%original: tensor<128x3xf32>) {
    ^bb0(%arg1: f32, %arg2: f32):
      %1 = addf %arg1, %arg2 : f32
      linalg_ext.yield %1 : f32
    } -> tensor<128x3xf32>
  return %0 : tensor<128x3xf32>
}
```

In this case, it's hard to flatten the flatten the indices being a scalar because you have to know the dim sizes of `originial` when lowering it to linalg_ext ops. We can probably create two dispatches, one is a linalg.generic op which would flatten the indices and reshape `original`, and another is a single `linalg_ext.scatter` op. But I think it does not worth doing this way.

In https://github.com/google/iree/pull/6413, we can trivially handle the case when lowering to loops.

```mlir
func @scatter_add_scalar_2D(
    %original: memref<4x3xi32>, %indices: memref<3x2xi32>,
    %updates: memref<3xi32>) {
  linalg_ext.scatter
    ins(%updates, %indices : memref<3xi32>, memref<3x2xi32>)
    outs(%original : memref<4x3xi32>)  {
  ^bb0(%arg0: i32, %arg1: i32):  // no predecessors
    %0 = addi %arg1, %arg0 : i32
    linalg_ext.yield %0 : i32
  }
  return
}
// CHECK-LABEL: func @scatter_add_scalar_2D
// CHECK-SAME:    %[[ORIGINAL:[a-zA-Z0-9]+]]
// CHECK-SAME:    %[[INDICES:[a-zA-Z0-9]+]]
// CHECK-SAME:    %[[UPDATES:[a-zA-Z0-9]+]]
// CHECK-DAG:     %[[C0:.+]] = constant 0 : index
// CHECK-DAG:     %[[C1:.+]] = constant 1 : index
// CHECK-DAG:     %[[C3:.+]] = constant 3 : index
// CHECK:         scf.for %[[I:.+]] = %[[C0]] to %[[C3]] step %[[C1]] {
// CHECK:           %[[T1:.+]] = memref.load %[[UPDATES]][%[[I]]] : memref<3xi32>
// CHECK:           %[[C0:.+]] = constant 0 : index
// CHECK:           %[[T2:.+]] = memref.load %[[INDICES]][%[[I]], %[[C0]]] : memref<3x2xi32>
// CHECK:           %[[IDX1:.+]] = index_cast %[[T2]] : i32 to index
// CHECK:           %[[C1:.+]] = constant 1 : index
// CHECK:           %[[T3:.+]] = memref.load %[[INDICES]][%[[I]], %[[C1]]] : memref<3x2xi32>
// CHECK:           %[[IDX2:.+]] = index_cast %[[T3]] : i32 to index
// CHECK:           %[[ORI:.+]] = memref.load %[[ORIGINAL]][%[[IDX1]], %[[IDX2]]] : memref<4x3xi32>
// CHECK:           %[[ADD:.+]] = addi %[[ORI]], %[[T1]] : i32
// CHECK:           memref.store %[[ADD]], %[[ORIGINAL]][%[[IDX1]], %[[IDX2]]]
```

See the snippet `%[[ORI:.+]] = memref.load %[[ORIGINAL]][%[[IDX1]], %[[IDX2]]] : memref<4x3xi32>`. We load the two indices and use them to load a value from `original`.

What do you think?

File: iree/compiler/Dialect/LinalgExt/IR/test/invalid.mlir [hunk 2]
```diff
@@ -27,11 +27,11 @@ func @sort_without_dimension(%arg0: tensor<3x4xi32>) -> tensor<3x4xi32> {
 // -----
 
 func @scatter_mixed_tensor_memref(
-    %update : memref<?x?xf32>, %indices : tensor<?xi32>,
+    %update : memref<?x?xf32>, %indices : tensor<?x1xi32>,
     %original : tensor<?x?xf32>) -> tensor<?x?xf32> {
   // expected-error @+1 {{expected `ins` operand #0 to be of RankedTensorType}}
   %0 = linalg_ext.scatter
-      ins(%update, %indices : memref<?x?xf32>, tensor<?xi32>)
+      ins(%update, %indices : memref<?x?xf32>, tensor<?x1xi32>)
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/LinalgExt/IR/test/invalid.mlir:34]: I am not sure why this is `tensor<?x1xi32>`. That is same as `tensor<?xi32>`. Maybe this should be canonicalized before being converted to the `linalg_ext.scatter` op

File: iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 3]
```diff
@@ -139,15 +145,18 @@ static LogicalResult verifyScatterOp(ScatterOp op) {
         "mismatch in shape of indices and update value at dim#0");
   }
   auto originalType = op.outputs()[0].getType().cast<ShapedType>();
-  if (originalType.getRank() != updateType.getRank()) {
+  // indexDepth + update dims should match to original dims. The first dim of
+  // update is the number of updates.
+  if (originalType.getRank() != indexDepth + updateType.getRank() - 1) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp:150]: This seems off to me as well. From my understanding
if the indices is of rank `M`, and original is of rank `N`, then updates is of rank `N - M + 1`.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Thanks Hanhan for offline clarifications. But the current implementation here seems to not match with what you described. I can chat more offline as well.

File: iree/compiler/Dialect/LinalgExt/IR/test/invalid.mlir (hunk 2 above)
Comment (hanhanW) [reply iree/compiler/Dialect/LinalgExt/IR/test/invalid.mlir:34]: As offline discussion, they are technically the same. Both are fine. Following the definition from tensorflow scatter op could be easier, like no need to determine if this is `index_detph=1` case.

https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update

**Review by MaheshRavishankar — APPROVED** [review]
Thanks Hanhan for the clarifications. Looks good.
